### DISCUSSION TITLE
Mutability of variational state parameters.

### DISCUSSION THREAD

**Discussion by skhaleefah** [netket/netket#1893] [Q&A] [answered]
Hi all, 

In this page [here](https://netket.readthedocs.io/en/latest/docs/varstate.html) under "Manipulating the parameters" it says that if you do not use `flax.core.copy()` you will only copy the outermost dictionaries but reference the inner ones and so modifying the inner ones would modify the original parameters as well. 

I am on netket version `3.11.2` and flax version `0.8.1` and find this to _not_ be the case. The screenshot demonstrates that there is no difference in referencing the parameters vs. copying them with `flax.core.copy()` when modifying the inner-most dictionaries. Is this supposed to be the case or am I misinterpreting something here? 

Thanks in advance!

![image](https://github.com/user-attachments/assets/69635d30-e3e2-4755-a5fd-4715aa71751d)

**Comment by PhilipVinc** [comment, marked as answer]
I think the documentation is correct **in general**.  See the following example

```python
In [18]: a = {'a': { 'b':1}}

In [19]: a_copy=a.copy()

In [20]: a['a']['b']=2

In [21]: a_copy
Out[21]: {'a': {'b': 2}}

In [22]: a_flax_copy=flax.core.copy(a, {})

In [23]: a['a']['b']=3

In [24]: a_flax_copy
Out[24]: {'a': {'b': 2}}
```

However, for the particular case you considered, NetKet is trying hard to ensure you do not modify the parameters within `vs.parameters`directly, and internally returns you a copy obtained with `flax.core.copy` of the parameters. 
See https://github.com/netket/netket/blob/c9c1061ff127f56f893b86f6d08406d57b4bcee5/netket/vqs/base.py#L71

This is because the variational state, to work correctly, requires that if you modify the parameters, you reassign them with `vs.parameters = new_parameters`. This is used in order to ensure that some caches (like the samples) are cleaned when you change parameters.
If you changed them directly by editing the dictionary, we would have no way of knowing, and therefore we could not clear the caches and samples automatically.
